ccm_web/server/src/api/api.create.section.handler.ts

```diff
@@ -0,0 +1,67 @@
+import CanvasRequestor from '@kth/canvas-api'
+
+import { CanvasSectionBase, CreateSectionsAPIErrorData, APIErrorData, isAPICreateSectionErrorData } from './api.interfaces'
+import { handleAPIError } from './api.utils'
+
+import baseLogger from '../logger'
+
+const logger = baseLogger.child({ filePath: __filename })
+
+export class CreateSectionApiHandler {
+  requestor: CanvasRequestor
+  sections: string[]
+  courseId: number
+
+  constructor (requestor: CanvasRequestor, sections: string[], courseId: number) {
+    this.requestor = requestor
+    this.sections = sections
+    this.courseId = courseId
+  }
+
+  async apiCreateSection (sectionName: string): Promise<CanvasSectionBase | CreateSectionsAPIErrorData> {
+    try {
+      const endpoint = `courses/${this.courseId}/sections`
+      const method = 'POST'
+      const requestBody = { course_section: { name: sectionName } }
+      logger.debug(`Sendings request to Canvas - Endpoint: ${endpoint}; Method: ${method}; Body: ${JSON.stringify(requestBody)}`)
+      const response = await this.requestor.requestUrl<CanvasSectionBase>(endpoint, method, requestBody)
+      const { id, name } = response.body
+      return { id, name }
+    } catch (error) {
+      const errResponse: APIErrorData = handleAPIError(error)
+      return { statusCode: errResponse.statusCode, errors: [{ message: errResponse.message, failedInput: sectionName }] }
+    }
+  }
+
+  makeReturnResponseCreateSections (sectionsReturnRes: Array<CreateSectionsAPIErrorData | CanvasSectionBase>): CanvasSectionBase[] | CreateSectionsAPIErrorData {
+    const successes = []; const statusCodes = []; const errorsList = []
+    for (const section of sectionsReturnRes) {
+      if (isAPICreateSectionErrorData(section)) {
+        const { statusCode, errors } = section
+        errorsList.push(...errors)
+        statusCodes.push(statusCode)
+      } else {
+        successes.push(section)
+      }
+    }
+
+    if (successes.length === this.sections.length) {
+      return successes
+    } else {
+      const uniqueStatusCodes = [...new Set(statusCodes)]
+      return {
+        statusCode: uniqueStatusCodes.length > 1 ? 400 : uniqueStatusCodes[0],
```

Comment:
Why a status code of 400 here?  That means "bad request".  What if the problem is not in the request to the CCM API, but is a problem with contacting the Canvas API?  In that case, giving a 400 here will give the impression to the frontend that its request was made incorrectly.  It hides the fact that something may have gone wrong in Canvas.

Reply:
What status code would you use then? This might need some additional discussion and testing.

Reply:
Yeah.... If there are 6 section needs to be created, probably we should expect that failure can occur for different reason, although we don't know how often we would hit this scenario. We could have a mix of [200, 400, 402, 503, 200, 200]

We can make it 500 making it as internal error, but we are logging all those status code in logs if we need to find more details.

We can discuss this in the meeting

Reply:
Can we use a custom code that is available as per HTTP standards? That way we know that when user reports failure and  give us a status code ( if they are that savvy) we can figure out this is custom error code https://stackoverflow.com/a/62866782

Thoughts???

Reply:
After the discussion on the slack and in meeting, I changed it to 502 as per @lsloan suggestion `Note that the definition of 502 seems a good fit for some problems CCM's API will handle with Canvas' API: https://datatracker.ietf.org/doc/html/rfc7231#section-6.6.3`